### ISSUE TITLE
PR 892 does not always work

### ISSUE THREAD

**Issue by mvorisek** [sebastianbergmann/php-code-coverage#932]
| Q                         | A
| --------------------------| ---------------
| php-code-coverage version | 9.2.17 (latest)
| PHP version               | 8.0.23 (latest)
| Driver                    | Xdebug
| Xdebug version  | 3.2.0alpha3 (latest)
| Installation Method       | Composer
| Usage Method              | PHPUnit & other
| PHPUnit version (if used) | 9.5.24 (latest)

An issue, that took me quite a while to figure out what is going on.

In atk4/ui we have the following coverage setup:

a) collect coverage in phpunit tests, config: https://github.com/atk4/ui/blob/33779a9414b624a8faa7e959883d4ba407099dcd/phpunit.xml.dist#L23-L29
b) collect coverage from web requests, config: https://github.com/atk4/ui/blob/33779a9414b624a8faa7e959883d4ba407099dcd/demos/init-app.php#L21-L24 and https://github.com/atk4/ui/blob/33779a9414b624a8faa7e959883d4ba407099dcd/src/Behat/CoverageUtil.php#L38-L43
c) then merge all the coverage files using phpunit/phpcov in https://github.com/atk4/ui/blob/33779a9414b624a8faa7e959883d4ba407099dcd/.github/workflows/test-unit.yml#L223

Coverage collection is working. But as soon as we include `tests/` dir also in b), the coverage decreases. This is wrong, as all `tests/` files are already included in a) and inclusion of an already included file in another coverage collection cannot decrease the merged coverage.

CI repro: https://github.com/atk4/ui/pull/1852/commits - see 2nd and 3rd commit

https://app.codecov.io/gh/atk4/ui/blob/b762d62434df393aad170a7378d3d45876f93f5a/tests/JsTest.php (2nd commit coverage)
https://app.codecov.io/gh/atk4/ui/blob/99f347b7d04567faefafa3b39105042b728e1649/tests/JsTest.php (3rd commit coverage)

It seems the problem is in https://github.com/sebastianbergmann/php-code-coverage/pull/892 or in https://github.com/sebastianbergmann/phpcov.

**Comment by sebastianbergmann** [comment]
CC @Slamdunk

**Comment by Slamdunk** [comment]
@mvorisek it seems `\SebastianBergmann\CodeCoverage\StaticAnalysis\ExecutableLinesFindingVisitor` doesn't get called at all, otherwise lines like `[10, 0]` would have never been registered as executable.

It's tough to debug just by looking at the code though: can you provide the complete `coverage/*.cov` files content?

**Comment by mvorisek** [comment]
> It's tough to debug just by looking at the code though: can you provide the complete `coverage/*.cov` files content?

@Slamdunk please see last 2 commits in https://github.com/atk4/ui/pull/1852/commits and related CI workflows:

- https://github.com/atk4/ui/actions/runs/2980968373
- https://github.com/atk4/ui/actions/runs/2980977387

there are 2 coverage CI pipelines per workflow - `unit` and `behat`, each uploads `merged.xml` file to CodeCov

for further analysis, please see the `artifact_behat_merged` and `artifact_behat_merged` artifacts, there are all generated coverage files + the `merged.xml`

**Comment by Slamdunk** [comment]
How do I download the artifacts?

**Comment by mvorisek** [comment]
![image](https://user-images.githubusercontent.com/2228672/188232800-a8f01c68-a653-4fc0-b200-21b06904dbd4.png)

just scroll down :)

**Comment by Slamdunk** [comment]
I don't see that secion, I think it's restricted to the owner of the repo.
Try visiting https://github.com/atk4/ui/actions/runs/2980968373 in browser's private mode

**Comment by mvorisek** [comment]
On my side, it is there also in browser's private mode:

![image](https://user-images.githubusercontent.com/2228672/188421998-20a83a80-f95e-46ec-81ff-4214a7d82811.png)

but the scrolling has some issue, try this:

- scroll manually by the rightmost browser window scroll bar (move it manually, not via mouse scroll button)
- or reduce the page zoom to like 50%, then you should see the artifacts /wo scroll, see my screenshot above

**Comment by Slamdunk** [comment]
`*.cov` files from artifacts seem to have an absolute path `/__w/ui/ui/*` which cannot interact properly with a `phpcov` local run; I am unable to debug what's happening on your CI.

My guess is that `ExecutableLinesFindingVisitor` doesn't get called at all because related files cannot be found https://github.com/sebastianbergmann/php-code-coverage/blob/aa94dc41e8661fe90c7316849907cba3007b10d8/src/CodeCoverage.php#L485-L487

But I need a much simpler setup to help you further

**Comment by mvorisek** [comment]
> `*.cov` files from artifacts seem to have an absolute path `/__w/ui/ui/*` which cannot interact properly with a `phpcov` local run; I am unable to debug what's happening on your CI.

we use pretty default config - https://github.com/atk4/ui/blob/33779a9414b624a8faa7e959883d4ba407099dcd/src/Behat/CoverageUtil.php#L28-L30

can we change something to change the output paths to relative?

or for the debug, can you clone https://github.com/atk4/ui/pull/1852/commits/4dc6e58e5c6e14c71d74313188a30f067d546ae9 and run phpunit locally?

all you should need is to:

1. clone the repo and cd to it
2. create `coverage` directory in it - https://github.com/atk4/ui/blob/4dc6e58e5c6e14c71d74313188a30f067d546ae9/.github/workflows/test-unit.yml#L123
3. setup php.ini for xdebug to collect coverage - https://github.com/atk4/ui/blob/4dc6e58e5c6e14c71d74313188a30f067d546ae9/.github/workflows/test-unit.yml#L95
4. create sqlite demo DB and run phpunit tests - https://github.com/atk4/ui/blob/4dc6e58e5c6e14c71d74313188a30f067d546ae9/.github/workflows/test-unit.yml#L127-L128
5. merge the coverage files - https://github.com/atk4/ui/blob/4dc6e58e5c6e14c71d74313188a30f067d546ae9/.github/workflows/test-unit.yml#L203

**Comment by Slamdunk** [comment]
What's the file I need to look for after merge?
Because the steps you provided produce an empty `JsTest.php` result:

![image](https://user-images.githubusercontent.com/152236/188447299-19f48c3a-04b2-4d45-8d65-a40a6b7c6aca.png)

**Comment by Slamdunk** [comment]
@mvorisek I got to the same conclusion.

@sebastianbergmann can you move this issue to https://github.com/sebastianbergmann/phpcov please?

**Comment by mvorisek** [comment]
If I am not mistaken, the problem is somewhere in https://github.com/sebastianbergmann/php-code-coverage/blob/5f9610ab5506c31c7035ba59d2da48408c50445b/src/CodeCoverage.php#L235 in this repo, phpcov repo seems to be only a CLI app wrapper

**Comment by Slamdunk** [comment]
I don't think so: I'm the maintainer of ParaTest where CC merging is a common practice (https://github.com/paratestphp/paratest/blob/c163539818fd96308ca8dc60f46088461e366ed4/src/Coverage/CoverageMerger.php#L36)
and never had this issue before.

https://github.com/sebastianbergmann/php-code-coverage/pull/892 born exactly from many many issues I had on parallel executions and CC merging

**Comment by Slamdunk** [comment]
@mvorisek btw, as soon as you are able to give us a reproducible local environment for the issue, I can debug it quickly. But without such premise, I can't help further

**Comment by mvorisek** [comment]
In our usecase, coverages from UNIT / BEHAT workflows are not merged. They are sent separately to CodeCov.

As analysed in https://github.com/sebastianbergmann/php-code-coverage/issues/932#issuecomment-1238304653, the problem is when uncovered file is merged (by this repo). The current result is zero file coverage /wo #892, ie. more uncovered lines than needed.

You should be able to demonstrate this problem in ParaTest too. Just add some file to your project that needs #892 but is never executed. You will get zero coverage correctly, but /w more uncovered lines than if at least 1 line would be executed.
